### ISSUE TITLE
Problems of the function metric_to_structure in TreeRep.jl?

### ISSUE THREAD

**Issue by elichienxD** [rsonthal/TreeRep#5]
Hi @rsonthal ,

Thanks for making the implementation of your interesting paper public available! I have two questions regarding the TreeRep algorithm.

In TreeRep.jl, line 17:

function metric_to_structure(d,p2,jj;increase = false, check_cluster = false)

What does p2 and jj use for? It is also not used in TreeRepy.py. I wonder if I can simply remove it?

On the other hand, I find that the current implementation of TreeRep (using python wrapper) fails to recover a simple binary tree with 4 leaves.

Suppose our input distance matrix is

[[0.  0.2 0.4 0.4]
 [0.2 0.  0.4 0.4]
 [0.4 0.4 0.  0.2]
 [0.4 0.4 0.2 0. ]]

This corresponds to the following balance binary tree

![image](https://user-images.githubusercontent.com/46387845/149613390-d3daf0a9-eea1-4116-b9ee-fcfe1ea00eae.png)

However, TreeRep (using python wrapper) will return the following tree

![image](https://user-images.githubusercontent.com/46387845/149613440-b593cb13-9051-4c9a-97e6-f42400cdceba.png)

 which correspond to the following distance matrix

[[0.   0.15 0.3  0.3 ]
 [0.15 0.   0.45 0.45]
 [0.3  0.45 0.   0.2 ]
 [0.3  0.45 0.2  0.  ]]

Is it an implementation problem? If not, it seems to contradict the Theorem 1 in the paper. Please let me know if I misunderstand something.

Thanks,
Eli

**Comment by rsonthal** [comment]
Hi @elichienxD, 

So p2 and jj where used to keep track of where it was in the algorithm when I was first debugging the implementation and are not currently needed, but you'll have to careful in getting rid of it (as it maybe referenced in various recursive function calls). 

I am not sure how that second graph corresponds to the second distance matrix. but the second picture is incorrect anyway.  I just tried running the Julia version of the code and got the following output and that looks correct to me? (node indices are off by 1, i.e. 0 -> 1, 1-> 2, ... and it switched/collapsed a Steiner node but it has the correct structure). 


<img width="702" alt="Screen Shot 2022-01-15 at 1 22 09 PM" src="https://user-images.githubusercontent.com/5423871/149614261-be0f8996-3d5b-417b-8067-c3948ed3b98f.png">

Below I am running the python wrapper (for full disclosure I had some trouble with the julia package in python, maybe https://github.com/nalexai/hyperlib has an easier to use wrapper). But the following tree looks correct to me (note this only has the edges) in the weight matrix. 

<img width="840" alt="Screen Shot 2022-01-15 at 1 42 33 PM" src="https://user-images.githubusercontent.com/5423871/149614782-f0c8d7f5-96c6-42c3-b194-a40860e0cfa3.png">

If you are still having trouble can you share the code you are running to see if I can reproduce the issue?

Thanks, 
Rishi

**Comment by rsonthal** [comment]
If I divide the weights by 10, then I can reproduce the issue. Let me look into this

**Comment by rsonthal** [comment]
The issue it seems was with the tol (tolerance) parameter. This parameter, says if distances are smaller than this, then collapse the two node. So by default this was set to 0.1, since minimum distances in my data sets were large. I have changed the default to 1e-5, and made tol a parameter that can be passed into the function. (I made the edits on GitHub itself and have tested it, so let me know if it works)

Thanks for noticing this!
